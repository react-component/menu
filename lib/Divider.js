'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var Divider = (0, _createReactClass2["default"])({
  displayName: 'Divider',

  propTypes: {
    disabled: _propTypes2["default"].bool,
    className: _propTypes2["default"].string,
    rootPrefixCls: _propTypes2["default"].string
  },

  getDefaultProps: function getDefaultProps() {
    return {
      disabled: true
    };
  },
  render: function render() {
    var _props = this.props,
        _props$className = _props.className,
        className = _props$className === undefined ? '' : _props$className,
        rootPrefixCls = _props.rootPrefixCls;

    return _react2["default"].createElement('li', { className: className + ' ' + rootPrefixCls + '-item-divider' });
  }
});

exports["default"] = Divider;
module.exports = exports['default'];