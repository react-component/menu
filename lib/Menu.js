'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _extends2 = require('babel-runtime/helpers/extends');

var _extends3 = _interopRequireDefault(_extends2);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _createReactClass = require('create-react-class');

var _createReactClass2 = _interopRequireDefault(_createReactClass);

var _MenuMixin = require('./MenuMixin');

var _MenuMixin2 = _interopRequireDefault(_MenuMixin);

var _util = require('./util');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

// import React from 'react';
var Menu = (0, _createReactClass2["default"])({
  displayName: 'Menu',

  propTypes: {
    openSubMenuOnMouseEnter: _propTypes2["default"].bool,
    closeSubMenuOnMouseLeave: _propTypes2["default"].bool,
    selectedKeys: _propTypes2["default"].arrayOf(_propTypes2["default"].string),
    defaultSelectedKeys: _propTypes2["default"].arrayOf(_propTypes2["default"].string),
    defaultOpenKeys: _propTypes2["default"].arrayOf(_propTypes2["default"].string),
    openKeys: _propTypes2["default"].arrayOf(_propTypes2["default"].string),
    mode: _propTypes2["default"].string,
    onClick: _propTypes2["default"].func,
    onSelect: _propTypes2["default"].func,
    onDeselect: _propTypes2["default"].func,
    onDestroy: _propTypes2["default"].func,
    openTransitionName: _propTypes2["default"].string,
    openAnimation: _propTypes2["default"].oneOfType([_propTypes2["default"].string, _propTypes2["default"].object]),
    level: _propTypes2["default"].number,
    eventKey: _propTypes2["default"].string,
    selectable: _propTypes2["default"].bool,
    children: _propTypes2["default"].any
  },

  mixins: [_MenuMixin2["default"]],

  getDefaultProps: function getDefaultProps() {
    return {
      openSubMenuOnMouseEnter: true,
      closeSubMenuOnMouseLeave: true,
      selectable: true,
      onClick: _util.noop,
      onSelect: _util.noop,
      onOpenChange: _util.noop,
      onDeselect: _util.noop,
      defaultSelectedKeys: [],
      defaultOpenKeys: []
    };
  },
  getInitialState: function getInitialState() {
    var props = this.props;
    var selectedKeys = props.defaultSelectedKeys;
    var openKeys = props.defaultOpenKeys;
    if ('selectedKeys' in props) {
      selectedKeys = props.selectedKeys || [];
    }
    if ('openKeys' in props) {
      openKeys = props.openKeys || [];
    }
    return {
      selectedKeys: selectedKeys,
      openKeys: openKeys
    };
  },
  componentWillReceiveProps: function componentWillReceiveProps(nextProps) {
    var props = {};
    if ('selectedKeys' in nextProps) {
      props.selectedKeys = nextProps.selectedKeys || [];
    }
    if ('openKeys' in nextProps) {
      props.openKeys = nextProps.openKeys || [];
    }
    this.setState(props);
  },
  onDestroy: function onDestroy(key) {
    var state = this.state;
    var props = this.props;
    var selectedKeys = state.selectedKeys;
    var openKeys = state.openKeys;
    var index = selectedKeys.indexOf(key);
    if (!('selectedKeys' in props) && index !== -1) {
      selectedKeys.splice(index, 1);
    }
    index = openKeys.indexOf(key);
    if (!('openKeys' in props) && index !== -1) {
      openKeys.splice(index, 1);
    }
  },
  onItemHover: function onItemHover(e) {
    var item = e.item;
    var _props = this.props,
        mode = _props.mode,
        closeSubMenuOnMouseLeave = _props.closeSubMenuOnMouseLeave;
    var _e$openChanges = e.openChanges,
        openChanges = _e$openChanges === undefined ? [] : _e$openChanges;
    // special for top sub menu

    if (mode !== 'inline' && !closeSubMenuOnMouseLeave && item.isSubMenu) {
      var activeKey = this.state.activeKey;
      var activeItem = this.getFlatInstanceArray().filter(function (c) {
        return c && c.props.eventKey === activeKey;
      })[0];
      if (activeItem && activeItem.props.open) {
        openChanges = openChanges.concat({
          key: item.props.eventKey,
          item: item,
          originalEvent: e,
          open: true
        });
      }
    }
    openChanges = openChanges.concat(this.getOpenChangesOnItemHover(e));
    if (openChanges.length) {
      this.onOpenChange(openChanges);
    }
  },
  onSelect: function onSelect(selectInfo) {
    var props = this.props;
    if (props.selectable) {
      // root menu
      var selectedKeys = this.state.selectedKeys;
      var selectedKey = selectInfo.key;
      if (props.multiple) {
        selectedKeys = selectedKeys.concat([selectedKey]);
      } else {
        selectedKeys = [selectedKey];
      }
      if (!('selectedKeys' in props)) {
        this.setState({
          selectedKeys: selectedKeys
        });
      }
      props.onSelect((0, _extends3["default"])({}, selectInfo, {
        selectedKeys: selectedKeys
      }));
    }
  },
  onClick: function onClick(e) {
    this.props.onClick(e);
  },
  onOpenChange: function onOpenChange(e_) {
    var props = this.props;
    var openKeys = this.state.openKeys.concat();
    var changed = false;
    var processSingle = function processSingle(e) {
      var oneChanged = false;
      if (e.open) {
        oneChanged = openKeys.indexOf(e.key) === -1;
        if (oneChanged) {
          openKeys.push(e.key);
        }
      } else {
        var index = openKeys.indexOf(e.key);
        oneChanged = index !== -1;
        if (oneChanged) {
          openKeys.splice(index, 1);
        }
      }
      changed = changed || oneChanged;
    };
    if (Array.isArray(e_)) {
      // batch change call
      e_.forEach(processSingle);
    } else {
      processSingle(e_);
    }
    if (changed) {
      if (!('openKeys' in this.props)) {
        this.setState({ openKeys: openKeys });
      }
      props.onOpenChange(openKeys);
    }
  },
  onDeselect: function onDeselect(selectInfo) {
    var props = this.props;
    if (props.selectable) {
      var selectedKeys = this.state.selectedKeys.concat();
      var selectedKey = selectInfo.key;
      var index = selectedKeys.indexOf(selectedKey);
      if (index !== -1) {
        selectedKeys.splice(index, 1);
      }
      if (!('selectedKeys' in props)) {
        this.setState({
          selectedKeys: selectedKeys
        });
      }
      props.onDeselect((0, _extends3["default"])({}, selectInfo, {
        selectedKeys: selectedKeys
      }));
    }
  },
  getOpenTransitionName: function getOpenTransitionName() {
    var props = this.props;
    var transitionName = props.openTransitionName;
    var animationName = props.openAnimation;
    if (!transitionName && typeof animationName === 'string') {
      transitionName = props.prefixCls + '-open-' + animationName;
    }
    return transitionName;
  },
  isInlineMode: function isInlineMode() {
    return this.props.mode === 'inline';
  },
  lastOpenSubMenu: function lastOpenSubMenu() {
    var lastOpen = [];
    var openKeys = this.state.openKeys;

    if (openKeys.length) {
      lastOpen = this.getFlatInstanceArray().filter(function (c) {
        return c && openKeys.indexOf(c.props.eventKey) !== -1;
      });
    }
    return lastOpen[0];
  },
  renderMenuItem: function renderMenuItem(c, i, subIndex) {
    if (!c) {
      return null;
    }
    var state = this.state;
    var extraProps = {
      openKeys: state.openKeys,
      selectedKeys: state.selectedKeys,
      openSubMenuOnMouseEnter: this.props.openSubMenuOnMouseEnter
    };
    return this.renderCommonMenuItem(c, i, subIndex, extraProps);
  },
  render: function render() {
    var props = (0, _extends3["default"])({}, this.props);
    props.className += ' ' + props.prefixCls + '-root';
    return this.renderRoot(props);
  }
});

exports["default"] = Menu;
module.exports = exports['default'];