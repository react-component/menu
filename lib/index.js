'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.Divider = exports.ItemGroup = exports.MenuItemGroup = exports.MenuItem = exports.Item = exports.SubMenu = undefined;

var _Menu = require('./Menu');

var _Menu2 = _interopRequireDefault(_Menu);

var _SubMenu = require('./SubMenu');

var _SubMenu2 = _interopRequireDefault(_SubMenu);

var _MenuItem = require('./MenuItem');

var _MenuItem2 = _interopRequireDefault(_MenuItem);

var _MenuItemGroup = require('./MenuItemGroup');

var _MenuItemGroup2 = _interopRequireDefault(_MenuItemGroup);

var _Divider = require('./Divider');

var _Divider2 = _interopRequireDefault(_Divider);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

exports.SubMenu = _SubMenu2["default"];
exports.Item = _MenuItem2["default"];
exports.MenuItem = _MenuItem2["default"];
exports.MenuItemGroup = _MenuItemGroup2["default"];
exports.ItemGroup = _MenuItemGroup2["default"];
exports.Divider = _Divider2["default"];
exports["default"] = _Menu2["default"];