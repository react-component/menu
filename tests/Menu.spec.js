/* eslint-disable no-undef, react/no-multi-comp, react/jsx-curly-brace-presence, max-classes-per-file */
import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import { resetWarned } from 'rc-util/lib/warning';
import Menu, { MenuItem, MenuItemGroup, SubMenu, Divider } from '../src';

describe('Menu', () => {
  describe('should render', () => {
    function createMenu(props) {
      return (
        <Menu
          disabledOverflow
          className="myMenu"
          openAnimation="fade"
          {...props}
        >
          <MenuItemGroup title="g1">
            <MenuItem key="1">1</MenuItem>
            <Divider />
            <MenuItem key="2">2</MenuItem>
          </MenuItemGroup>
          <MenuItem key="3">3</MenuItem>
          <MenuItemGroup title="g2">
            <MenuItem key="4">4</MenuItem>
            <MenuItem key="5" disabled>
              5
            </MenuItem>
          </MenuItemGroup>
          <SubMenu title="submenu">
            <MenuItem key="6">6</MenuItem>
          </SubMenu>
        </Menu>
      );
    }

    ['vertical', 'horizontal', 'inline'].forEach(mode => {
      it(`${mode} menu correctly`, () => {
        const wrapper = mount(createMenu({ mode }));
        expect(wrapper.render()).toMatchSnapshot();
      });

      it(`${mode} menu with empty children without error`, () => {
        expect(() => mount(<Menu mode={mode}>{[]}</Menu>)).not.toThrow();
      });

      it(`${mode} menu with undefined children without error`, () => {
        expect(() => mount(<Menu mode={mode} />)).not.toThrow();
      });

      it(`${mode} menu that has a submenu with undefined children without error`, () => {
        expect(() =>
          mount(
            <Menu mode={mode}>
              <SubMenu />
            </Menu>,
          ),
        ).not.toThrow();
      });

      it(`${mode} menu with rtl direction correctly`, () => {
        const wrapper = mount(createMenu({ mode, direction: 'rtl' }));
        expect(wrapper.render()).toMatchSnapshot();

        expect(wrapper.find('ul').first().props().className).toContain('-rtl');
      });
    });

    it('should support Fragment', () => {
      const wrapper = mount(
        <Menu>
          <SubMenu title="submenu">
            <MenuItem key="6">6</MenuItem>
          </SubMenu>
          <MenuItem key="7">6</MenuItem>
          <>
            <SubMenu title="submenu">
              <MenuItem key="8">6</MenuItem>
            </SubMenu>
            <MenuItem key="9">6</MenuItem>
          </>
        </Menu>,
      );
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  describe('render role listbox', () => {
    function createMenu() {
      return (
        <Menu className="myMenu" openAnimation="fade" role="listbox">
          <MenuItem key="1" role="option">
            1
          </MenuItem>
          <MenuItem key="2" role="option">
            2
          </MenuItem>
          <MenuItem key="3" role="option">
            3
          </MenuItem>
        </Menu>
      );
    }

    it('renders menu correctly', () => {
      const wrapper = mount(createMenu());
      expect(wrapper.render()).toMatchSnapshot();
    });
  });

  it('set activeKey', () => {
    const wrapper = mount(
      <Menu activeKey="1">
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
      </Menu>,
    );
    expect(wrapper.isActive(0)).toBeTruthy();
    expect(wrapper.isActive(1)).toBeFalsy();

    wrapper.setProps({ activeKey: '2' });
    expect(wrapper.isActive(0)).toBeFalsy();
    expect(wrapper.isActive(1)).toBeTruthy();
  });

  it('active first item', () => {
    const wrapper = mount(
      <Menu defaultActiveFirst>
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
      </Menu>,
    );
    expect(
      wrapper.find('.rc-menu-item').first().hasClass('rc-menu-item-active'),
    ).toBeTruthy();
  });

  it('should render none menu item children', () => {
    expect(() => {
      mount(
        <Menu activeKey="1">
          <MenuItem key="1">1</MenuItem>
          <MenuItem key="2">2</MenuItem>
          string
          {'string'}
          {null}
          {undefined}
          {12345}
          <div />
          <input />
        </Menu>,
      );
    }).not.toThrow();
  });

  it('select multiple items', () => {
    const wrapper = mount(
      <Menu multiple>
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
      </Menu>,
    );
    wrapper.find('.rc-menu-item').first().simulate('click');
    wrapper.find('.rc-menu-item').last().simulate('click');

    expect(wrapper.find('li.rc-menu-item-selected')).toHaveLength(2);
  });

  it('can be controlled by selectedKeys', () => {
    const wrapper = mount(
      <Menu selectedKeys={['1']}>
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
      </Menu>,
    );
    expect(wrapper.find('li').first().props().className).toContain('-selected');
    wrapper.setProps({ selectedKeys: ['2'] });
    wrapper.update();
    expect(wrapper.find('li').last().props().className).toContain('-selected');
  });

  it('empty selectedKeys not to throw', () => {
    mount(
      <Menu selectedKeys={null}>
        <MenuItem key="foo">foo</MenuItem>
      </Menu>,
    );
  });

  it('not selectable', () => {
    const onSelect = jest.fn();

    const wrapper = mount(
      <Menu onSelect={onSelect} selectedKeys={[]}>
        <MenuItem key="bamboo">Bamboo</MenuItem>
      </Menu>,
    );

    wrapper.findItem(0).simulate('click');
    expect(onSelect).toHaveBeenCalledWith(
      expect.objectContaining({ selectedKeys: ['bamboo'] }),
    );

    onSelect.mockReset();
    wrapper.setProps({ selectable: false });
    wrapper.findItem(0).simulate('click');
    expect(onSelect).not.toHaveBeenCalled();
  });

  it('select default item', () => {
    const wrapper = mount(
      <Menu defaultSelectedKeys={['1']}>
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
      </Menu>,
    );
    expect(wrapper.find('li').first().props().className).toContain('-selected');
  });

  it('issue https://github.com/ant-design/ant-design/issues/29429', () => {
    // don't use selectedKeys as string
    // it is a compatible feature for https://github.com/ant-design/ant-design/issues/29429
    const wrapper = mount(
      <Menu selectedKeys="item_abc">
        <MenuItem key="item_a">1</MenuItem>
        <MenuItem key="item_abc">2</MenuItem>
      </Menu>,
    );
    expect(wrapper.find('li').at(0).props().className).not.toContain(
      '-selected',
    );
    expect(wrapper.find('li').at(1).props().className).toContain('-selected');
  });

  describe('openKeys', () => {
    it('can be controlled by openKeys', () => {
      const wrapper = mount(
        <Menu openKeys={['g1']}>
          <Menu.SubMenu key="g1">
            <MenuItem key="1">1</MenuItem>
          </Menu.SubMenu>
          <Menu.SubMenu key="g2">
            <MenuItem key="2">2</MenuItem>
          </Menu.SubMenu>
        </Menu>,
      );

      expect(
        wrapper.find('InlineSubMenuList').first().props().open,
      ).toBeTruthy();
      expect(wrapper.find('InlineSubMenuList').last().props().open).toBeFalsy();

      wrapper.setProps({ openKeys: ['g2'] });
      expect(
        wrapper.find('InlineSubMenuList').first().props().open,
      ).toBeFalsy();
      expect(
        wrapper.find('InlineSubMenuList').last().props().open,
      ).toBeTruthy();
    });

    it('openKeys should allow to be empty', () => {
      const wrapper = mount(
        <Menu
          onClick={() => {}}
          onOpenChange={() => {}}
          openKeys={undefined}
          selectedKeys={['1']}
          mode="inline"
        >
          <SubMenu title="1231">
            <MenuItem>
              <a>
                <span>123123</span>
              </a>
            </MenuItem>
          </SubMenu>
        </Menu>,
      );
      expect(wrapper).toBeTruthy();
    });

    it('null of openKeys', () => {
      const wrapper = mount(
        <Menu openKeys={null} mode="inline">
          <SubMenu key="bamboo" title="Bamboo">
            <MenuItem key="light">Light</MenuItem>
          </SubMenu>
        </Menu>,
      );
      expect(wrapper).toBeTruthy();
    });
  });

  it('open default submenu', () => {
    jest.useFakeTimers();

    const wrapper = mount(
      <Menu defaultOpenKeys={['g1']}>
        <SubMenu key="g1">
          <MenuItem key="1">1</MenuItem>
        </SubMenu>
        <SubMenu key="g2">
          <MenuItem key="2">2</MenuItem>
        </SubMenu>
      </Menu>,
    );

    act(() => {
      jest.runAllTimers();
      wrapper.update();
    });

    expect(wrapper.find('PopupTrigger').first().props().visible).toBeTruthy();
    expect(wrapper.find('PopupTrigger').last().props().visible).toBeFalsy();

    jest.useRealTimers();
  });

  it('fires select event', () => {
    const handleSelect = jest.fn();
    const wrapper = mount(
      <Menu onSelect={handleSelect}>
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
      </Menu>,
    );
    wrapper.find('MenuItem').first().simulate('click');
    expect(handleSelect.mock.calls[0][0].key).toBe('1');
  });

  it('fires click event', () => {
    jest.useFakeTimers();

    resetWarned();

    const errorSpy = jest.spyOn(console, 'error');

    const handleClick = jest.fn();
    const wrapper = mount(
      <Menu onClick={handleClick} openKeys={['parent']}>
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
        <Menu.SubMenu key="parent">
          <MenuItem key="3">3</MenuItem>
        </Menu.SubMenu>
      </Menu>,
    );

    act(() => {
      jest.runAllTimers();
      wrapper.update();
    });

    wrapper.find('.rc-menu-item').first().simulate('click');
    const info = handleClick.mock.calls[0][0];
    expect(info.key).toBe('1');
    expect(info.item).toBeTruthy();

    expect(errorSpy).toHaveBeenCalledWith(
      'Warning: `info.item` is deprecated since we will move to function component that not provides React Node instance in future.',
    );

    handleClick.mockReset();
    wrapper.find('.rc-menu-item').last().simulate('click');
    expect(handleClick.mock.calls[0][0].keyPath).toEqual(['3', 'parent']);

    errorSpy.mockRestore();

    jest.useRealTimers();
  });

  it('fires deselect event', () => {
    const handleDeselect = jest.fn();
    const wrapper = mount(
      <Menu multiple onDeselect={handleDeselect}>
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
      </Menu>,
    );
    wrapper.find('MenuItem').first().simulate('click').simulate('click');
    expect(handleDeselect.mock.calls[0][0].key).toBe('1');
  });

  it('active by mouse enter', () => {
    const wrapper = mount(
      <Menu>
        <MenuItem key="item1">item</MenuItem>
        <MenuItem disabled>disabled</MenuItem>
        <MenuItem key="item2">item2</MenuItem>
      </Menu>,
    );
    wrapper.find('li').last().simulate('mouseEnter');
    expect(wrapper.isActive(2)).toBeTruthy();
  });

  it('active by key down', () => {
    const wrapper = mount(
      <Menu activeKey="1">
        <MenuItem key="1">1</MenuItem>
        <MenuItem key="2">2</MenuItem>
      </Menu>,
    );

    // KeyDown will not change activeKey since control
    wrapper.find('Overflow').simulate('keyDown', { which: KeyCode.DOWN });
    expect(wrapper.isActive(0)).toBeTruthy();

    wrapper.setProps({ activeKey: '2' });
    expect(wrapper.isActive(1)).toBeTruthy();
  });

  it('defaultActiveFirst', () => {
    const wrapper = mount(
      <Menu selectedKeys={['foo']} defaultActiveFirst>
        <MenuItem key="foo">foo</MenuItem>
      </Menu>,
    );
    expect(wrapper.isActive(0)).toBeTruthy();
  });

  it('should accept builtinPlacements', () => {
    const builtinPlacements = {
      leftTop: {
        points: ['tr', 'tl'],
        overflow: {
          adjustX: 0,
          adjustY: 0,
        },
        offset: [0, 0],
      },
    };

    const wrapper = mount(
      <Menu builtinPlacements={builtinPlacements}>
        <MenuItem>menuItem</MenuItem>
        <SubMenu title="submenu">
          <MenuItem>menuItem</MenuItem>
        </SubMenu>
      </Menu>,
    );

    expect(wrapper.find('Trigger').prop('builtinPlacements').leftTop).toEqual(
      builtinPlacements.leftTop,
    );
  });

  describe('motion', () => {
    const defaultMotions = {
      inline: { motionName: 'inlineMotion' },
      horizontal: { motionName: 'horizontalMotion' },
      other: { motionName: 'defaultMotion' },
    };

    it('defaultMotions should work correctly', () => {
      const wrapper = mount(
        <Menu mode="inline" defaultMotions={defaultMotions}>
          <SubMenu key="bamboo">
            <MenuItem key="light" />
          </SubMenu>
        </Menu>,
      );

      // Inline
      wrapper.setProps({ mode: 'inline' });
      expect(wrapper.find('CSSMotion').last().prop('motionName')).toEqual(
        'inlineMotion',
      );

      // Horizontal
      wrapper.setProps({ mode: 'horizontal' });
      expect(
        wrapper.find('Trigger').last().prop('popupMotion').motionName,
      ).toEqual('horizontalMotion');

      // Default
      wrapper.setProps({ mode: 'vertical' });
      expect(
        wrapper.find('Trigger').last().prop('popupMotion').motionName,
      ).toEqual('defaultMotion');
    });

    it('motion is first level', () => {
      const wrapper = mount(
        <Menu
          mode="inline"
          defaultMotions={defaultMotions}
          motion={{ motionName: 'bambooLight' }}
        >
          <SubMenu key="bamboo">
            <MenuItem key="light" />
          </SubMenu>
        </Menu>,
      );

      // Inline
      wrapper.setProps({ mode: 'inline' });
      expect(wrapper.find('CSSMotion').last().prop('motionName')).toEqual(
        'bambooLight',
      );

      // Horizontal
      wrapper.setProps({ mode: 'horizontal' });
      expect(
        wrapper.find('Trigger').last().prop('popupMotion').motionName,
      ).toEqual('bambooLight');

      // Default
      wrapper.setProps({ mode: 'vertical' });
      expect(
        wrapper.find('Trigger').last().prop('popupMotion').motionName,
      ).toEqual('bambooLight');
    });
  });

  it('onMouseEnter should work', () => {
    const onMouseEnter = jest.fn();
    const wrapper = mount(
      <Menu onMouseEnter={onMouseEnter} defaultSelectedKeys={['test1']}>
        <MenuItem key="test1">Navigation One</MenuItem>
        <MenuItem key="test2">Navigation Two</MenuItem>
      </Menu>,
    );

    wrapper.find('ul.rc-menu-root').simulate('mouseEnter');
    expect(onMouseEnter).toHaveBeenCalled();
  });

  it('Nest children active should bump to top', async () => {
    const wrapper = mount(
      <Menu activeKey="light" mode="vertical">
        <SubMenu key="bamboo" title="Bamboo">
          <MenuItem key="light">Light</MenuItem>
        </SubMenu>
      </Menu>,
    );

    expect(wrapper.exists('.rc-menu-submenu-active')).toBeTruthy();
  });

  it('not warning on destroy', async () => {
    const errorSpy = jest.spyOn(console, 'error');

    const wrapper = mount(
      <Menu>
        <MenuItem key="bamboo">Bamboo</MenuItem>
      </Menu>,
    );

    wrapper.unmount();

    await Promise.resolve();

    expect(errorSpy).not.toHaveBeenCalled();

    errorSpy.mockRestore();
  });

  describe('Click should close Menu', () => {
    function test(name, props) {
      it(name, async () => {
        jest.useFakeTimers();

        const onOpenChange = jest.fn();

        const wrapper = mount(
          <Menu
            openKeys={['bamboo']}
            mode="vertical"
            onOpenChange={onOpenChange}
            {...props}
          >
            <SubMenu key="bamboo" title="Bamboo">
              <MenuItem key="light">Light</MenuItem>
            </SubMenu>
          </Menu>,
        );

        // Open menu
        await act(async () => {
          jest.runAllTimers();
          wrapper.update();
        });

        wrapper.update();

        wrapper.find('.rc-menu-item').last().simulate('click');
        expect(onOpenChange).toHaveBeenCalledWith([]);

        jest.useRealTimers();
      });
    }

    test('basic');
    test('not selectable', { selectable: false });
    test('inlineCollapsed', { mode: 'inline', inlineCollapsed: true });

    it('not close inline', async () => {
      jest.useFakeTimers();

      const onOpenChange = jest.fn();

      const wrapper = mount(
        <Menu openKeys={['bamboo']} mode="inline" onOpenChange={onOpenChange}>
          <SubMenu key="bamboo" title="Bamboo">
            <MenuItem key="light">Light</MenuItem>
          </SubMenu>
        </Menu>,
      );

      // Open menu
      await act(async () => {
        jest.runAllTimers();
        wrapper.update();
      });

      wrapper.update();

      wrapper.find('.rc-menu-item').last().simulate('click');
      expect(onOpenChange).not.toHaveBeenCalled();

      jest.useRealTimers();
    });
  });
});
/* eslint-enable */
