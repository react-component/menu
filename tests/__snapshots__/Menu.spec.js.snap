// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu render renders horizontal menu correctly 1`] = `
<ul
  class="rc-menu myMenu rc-menu-root rc-menu-horizontal"
  role="menu"
  style="position:relative"
  tabindex="0"
>
  <li
    class=" rc-menu-item-group"
  >
    <div
      class="rc-menu-item-group-title"
      title="g1"
    >
      g1
    </div>
    <ul
      class="rc-menu-item-group-list"
    >
      <li
        class="rc-menu-item"
        role="menuitem"
      >
        1
      </li>
      <li
        class=" rc-menu-item-divider"
      />
      <li
        class="rc-menu-item"
        role="menuitem"
      >
        2
      </li>
    </ul>
  </li>
  <li
    class="rc-menu-item"
    role="menuitem"
  >
    3
  </li>
  <li
    class=" rc-menu-item-group"
  >
    <div
      class="rc-menu-item-group-title"
      title="g2"
    >
      g2
    </div>
    <ul
      class="rc-menu-item-group-list"
    >
      <li
        class="rc-menu-item"
        role="menuitem"
      >
        4
      </li>
      <li
        aria-disabled="true"
        class="rc-menu-item rc-menu-item-disabled"
        role="menuitem"
      >
        5
      </li>
    </ul>
  </li>
  <li
    class="rc-menu-submenu rc-menu-submenu-horizontal"
    role="menuitem"
  >
    <div
      aria-expanded="false"
      aria-haspopup="true"
      class="rc-menu-submenu-title"
      title="submenu"
    >
      submenu
      <i
        class="rc-menu-submenu-arrow"
      />
    </div>
  </li>
  <li
    class="rc-menu-submenu rc-menu-submenu-horizontal rc-menu-overflowed-submenu"
    role="menuitem"
    style="position:absolute;left:0;visibility:hidden;height:0"
  >
    <div
      aria-expanded="false"
      aria-haspopup="true"
      class="rc-menu-submenu-title"
    >
      <span>
        ···
      </span>
      <i
        class="rc-menu-submenu-arrow"
      />
    </div>
  </li>
</ul>
`;

exports[`Menu render renders inline menu correctly 1`] = `
<ul
  class="rc-menu myMenu rc-menu-root rc-menu-inline"
  role="menu"
  style="position:relative"
  tabindex="0"
>
  <li
    class=" rc-menu-item-group"
  >
    <div
      class="rc-menu-item-group-title"
      title="g1"
    >
      g1
    </div>
    <ul
      class="rc-menu-item-group-list"
    >
      <li
        class="rc-menu-item"
        role="menuitem"
        style="padding-left:24px"
      >
        1
      </li>
      <li
        class=" rc-menu-item-divider"
      />
      <li
        class="rc-menu-item"
        role="menuitem"
        style="padding-left:24px"
      >
        2
      </li>
    </ul>
  </li>
  <li
    class="rc-menu-item"
    role="menuitem"
    style="padding-left:24px"
  >
    3
  </li>
  <li
    class=" rc-menu-item-group"
  >
    <div
      class="rc-menu-item-group-title"
      title="g2"
    >
      g2
    </div>
    <ul
      class="rc-menu-item-group-list"
    >
      <li
        class="rc-menu-item"
        role="menuitem"
        style="padding-left:24px"
      >
        4
      </li>
      <li
        aria-disabled="true"
        class="rc-menu-item rc-menu-item-disabled"
        role="menuitem"
        style="padding-left:24px"
      >
        5
      </li>
    </ul>
  </li>
  <li
    class="rc-menu-submenu rc-menu-submenu-inline"
    role="menuitem"
  >
    <div
      aria-expanded="false"
      aria-haspopup="true"
      class="rc-menu-submenu-title"
      style="padding-left:24px"
      title="submenu"
    >
      submenu
      <i
        class="rc-menu-submenu-arrow"
      />
    </div>
    <div />
  </li>
</ul>
`;

exports[`Menu render renders vertical menu correctly 1`] = `
<ul
  class="rc-menu myMenu rc-menu-root rc-menu-vertical"
  role="menu"
  style="position:relative"
  tabindex="0"
>
  <li
    class=" rc-menu-item-group"
  >
    <div
      class="rc-menu-item-group-title"
      title="g1"
    >
      g1
    </div>
    <ul
      class="rc-menu-item-group-list"
    >
      <li
        class="rc-menu-item"
        role="menuitem"
      >
        1
      </li>
      <li
        class=" rc-menu-item-divider"
      />
      <li
        class="rc-menu-item"
        role="menuitem"
      >
        2
      </li>
    </ul>
  </li>
  <li
    class="rc-menu-item"
    role="menuitem"
  >
    3
  </li>
  <li
    class=" rc-menu-item-group"
  >
    <div
      class="rc-menu-item-group-title"
      title="g2"
    >
      g2
    </div>
    <ul
      class="rc-menu-item-group-list"
    >
      <li
        class="rc-menu-item"
        role="menuitem"
      >
        4
      </li>
      <li
        aria-disabled="true"
        class="rc-menu-item rc-menu-item-disabled"
        role="menuitem"
      >
        5
      </li>
    </ul>
  </li>
  <li
    class="rc-menu-submenu rc-menu-submenu-vertical"
    role="menuitem"
  >
    <div
      aria-expanded="false"
      aria-haspopup="true"
      class="rc-menu-submenu-title"
      title="submenu"
    >
      submenu
      <i
        class="rc-menu-submenu-arrow"
      />
    </div>
  </li>
</ul>
`;

exports[`Menu render role listbox renders menu correctly 1`] = `
<ul
  class="rc-menu myMenu rc-menu-root rc-menu-vertical"
  role="listbox"
  style="position:relative"
  tabindex="0"
>
  <li
    aria-selected="false"
    class="rc-menu-item"
    role="option"
  >
    1
  </li>
  <li
    aria-selected="false"
    class="rc-menu-item"
    role="option"
  >
    2
  </li>
  <li
    aria-selected="false"
    class="rc-menu-item"
    role="option"
  >
    3
  </li>
</ul>
`;
